File: native/src/CaptureService.cpp
#include "CaptureService.h"

#include <QBuffer>
#include <QCoreApplication>
#include <QCryptographicHash>
#include <QGuiApplication>
#include <QPixmap>
#include <QScreen>
#include <QThread>
#include <algorithm>
#include <cmath>
#include <stdexcept>

QImage CaptureService::capturePrimaryMonitor()
{
    QCoreApplication *application = QGuiApplication::instance();
    if (!application || QThread::currentThread() != application->thread()) {
        throw std::runtime_error("Primary monitor capture must run on the Qt GUI thread.");
    }
    QScreen *screen = QGuiApplication::primaryScreen();
    if (!screen) {
        throw std::runtime_error("Could not open primary monitor.");
    }
    QPixmap pixmap = screen->grabWindow(0);
    if (pixmap.isNull()) {
        throw std::runtime_error("Primary monitor capture returned an empty image.");
    }
    return pixmap.toImage().convertToFormat(QImage::Format_RGB888);
}

QImage CaptureService::resizePreservingAspect(const QImage &image, int maxEdge)
{
    if (maxEdge <= 0) {
        throw std::invalid_argument("max_edge must be positive");
    }
    const QImage rgb = image.convertToFormat(QImage::Format_RGB888);
    const int longest = std::max(rgb.width(), rgb.height());
    if (longest <= maxEdge) {
        return rgb;
    }
    const double scale = static_cast<double>(maxEdge) / static_cast<double>(longest);
    const QSize size(std::max(1, static_cast<int>(std::round(rgb.width() * scale))),
                     std::max(1, static_cast<int>(std::round(rgb.height() * scale))));
    return rgb.scaled(size, Qt::KeepAspectRatio, Qt::SmoothTransformation).convertToFormat(QImage::Format_RGB888);
}

QString CaptureService::encodeJpegBase64(const QImage &image, int maxEdge, int quality)
{
    const QImage resized = resizePreservingAspect(image, maxEdge);
    QByteArray bytes;
    QBuffer buffer(&bytes);
    buffer.open(QIODevice::WriteOnly);
    resized.save(&buffer, "JPEG", quality);
    return QString::fromLatin1(bytes.toBase64());
}

QString CaptureService::imageFingerprint(const QImage &image)
{
    const QImage rgb = image.convertToFormat(QImage::Format_RGB888);
    QCryptographicHash hash(QCryptographicHash::Sha256);
    for (int y = 0; y < rgb.height(); ++y) {
        hash.addData(reinterpret_cast<const char *>(rgb.constScanLine(y)), rgb.width() * 3);
    }
    return QString::fromLatin1(hash.result().toHex().left(10));
}

File: src/ollama_hud/services/capture_service.py
from __future__ import annotations

import base64
import hashlib
import io

from PIL import Image


def capture_primary_monitor() -> Image.Image:
    try:
        import dxcam
    except ModuleNotFoundError as exc:
        raise RuntimeError("Primary monitor capture requires dxcam on Windows.") from exc

    camera = dxcam.create(output_idx=0)
    if camera is None:
        raise RuntimeError("Could not open primary monitor with dxcam.")
    try:
        frame = camera.grab()
    finally:
        camera.release()
    if frame is None:
        raise RuntimeError("dxcam did not return a frame.")
    return Image.fromarray(frame).convert("RGB")


def resize_preserving_aspect(image: Image.Image, max_edge: int) -> Image.Image:
    if max_edge <= 0:
        raise ValueError("max_edge must be positive")
    width, height = image.size
    longest = max(width, height)
    if longest <= max_edge:
        return image.convert("RGB")
    scale = max_edge / longest
    size = (max(1, round(width * scale)), max(1, round(height * scale)))
    return image.convert("RGB").resize(size, Image.Resampling.LANCZOS)


def encode_jpeg_base64(image: Image.Image, *, max_edge: int, quality: int) -> str:
    resized = resize_preserving_aspect(image, max_edge)
    buffer = io.BytesIO()
    resized.save(buffer, format="JPEG", quality=quality, optimize=True)
    return base64.b64encode(buffer.getvalue()).decode("ascii")


def image_fingerprint(image: Image.Image) -> str:
    rgb = image.convert("RGB")
    digest = hashlib.sha256(rgb.tobytes()).hexdigest()
    return digest[:10]
